### ISSUE TITLE
_ONEDPL_PRAGMA_UNROLL is undefined on open-source DPC++

### ISSUE THREAD

**Issue by joeatodd** [oneapi-src/oneDPL#1294]
While investigating performance of our new scan implementation, I found that `_ONEDPL_PRAGMA_UNROLL` is undefined for open-source DPC++.

I am using a recent build of the open-source compiler:
`clang version 18.0.0 (git@github.com:intel/llvm.git 05e7395ebe5333048b851dea608260b51cec5f69)`

I suspect this is because the reported GCC version is lower than it ought to be, as per this comment in `onedpl_config.h`:
```cpp
// note that when ICC or Clang is in use, _ONEDPL_GCC_VERSION might not fully match
// the actual GCC version on the system.
#define _ONEDPL_GCC_VERSION (__GNUC__ * 10000 + __GNUC_MINOR__ * 100 + __GNUC_PATCHLEVEL__)
```

If I run `echo | clang++ -fsycl -dM -E -x c++ - | grep -i '__GNUC'` on my machine I get:
```cpp
#define __GNUC_GNU_INLINE__ 1
#define __GNUC_MINOR__ 2
#define __GNUC_PATCHLEVEL__ 1
#define __GNUC__ 4
```
which means that this:
```cpp
// Enable loop unrolling pragmas where supported
#if (__INTEL_COMPILER || (!defined(__INTEL_COMPILER) && _ONEDPL_GCC_VERSION >= 80000))
#    define _ONEDPL_PRAGMA_UNROLL _ONEDPL_PRAGMA(unroll)
#else //no pragma unroll
#    define _ONEDPL_PRAGMA_UNROLL
#endif
```
makes `_ONEDPL_PRAGMA_UNROLL` undefined.

I also confirmed this empirically by replacing `_ONEDPL_PRAGMA_UNROLL` with `#pragma unroll` in my kernel and I got a consistent speedup.

**Comment by joeatodd** [comment]
Could we use `_ONEDPL_CLANG_VERSION` to fix this? Does anyone know which was the first version of `clang` would be the earliest which supports pragma unroll?

**Comment by joeatodd** [comment]
I see [here](https://releases.llvm.org/4.0.0/tools/clang/docs/AttributeReference.html#pragma-unroll-pragma-nounroll) that clang++ has supported `#pragma unroll` at least since 4.0. [This stackoverflow answer](https://stackoverflow.com/questions/27310229/loop-unrolling-in-clang) implies there was no support in 3.5, and that in 3.6 the syntax was very different.

So possibly:
```c++
#if (__INTEL_COMPILER || (!defined(__INTEL_COMPILER) && ((_ONEDPL_GCC_VERSION >= 80000) || (_ONEDPL_CLANG_VERSION >= 40000)))
#    define _ONEDPL_PRAGMA_UNROLL _ONEDPL_PRAGMA(unroll)
#else //no pragma unroll
#    define _ONEDPL_PRAGMA_UNROLL
#endif
```

**Comment by jinz2014** [comment]
Could a warning message about "no program unroll" be produced when it is undefined ?

**Comment by joeatodd** [comment]
Fixed by #1299
